docs/virtualpage.rst

```diff
@@ -0,0 +1,293 @@
+also makes it possible for VirtualPage to be embedded into any :ref:`component` quite reliably.
+
+To illustrate, how :php:class:`Tabs` component rely on VirtualPage, the following code::
+
+    $t = $layout->add('Tabs');
+
+    $t->addTab('Tab1')->add('LoremIpsum'); // regular tab
+    $t->addTab('Tab2', function($p){ $p->add('LoremIpsum'); }); // dynamic tab
+
+uses the following implementation of addTab()::
+
+    public function addTab($name = null, $action = null)
+    {
+        // add tabs menu item
+        $item = $this->add([$item, 'class'=>['item']], 'Menu');
+        $item->setElement('a');
+        $item->setAttr('data-tab', $item->name);
+
+        // add tabs sub-view
+        $sub = $this->add(['View', 'class'=>['ui tab']], 'Tabs');
+        $sub->setAttr('data-tab', $item->name);
+
+        // if there is callback action, then
+        if ($action && is_callable($action)) {
+            $vp = $sub->add('VirtualPage');
+            $item->setPath($vp->getUrl());
+
+            $vp->set($action);
+        }
+
+        return  $sub;
+    }
+
+You do not need to fully understand the code to notice that Dynamic Content support requires just
+few lines of extra code.
+
+.. php:method:: getURL($html_wrapping)
+
+    You can use this shortcut method instead of $vp->cb->getURL().
+
+.. php:attr:: $ui
+
+When using 'popup' mode, the output appears inside a `<div class="ui container">`. If you want to change this
+class, you can set $ui property to something else. Try::
+
+    $vp = $layout->add('VirtualPage');
+    $vp->add('LoremIpsum');
+    $vp->ui = 'red inverted segment';
+
+    $label = $layout->add('Label');
+
+    $label->detail = $vp->cb->getURL('popup');
+    $label->link($vp->cb->getURL('popup'));
+
+
+
+
+
+Loader
+------
+
+.. php:class:: Loader
+
+.. php:method:: set()
+
+Loader is designed to delay some slow-loading content by loading it dynamically, after main
+page is rendered.
+
+Comparing to VirtualPage which is a D.Y.I. solution - Loader can be used out of the box.
+Loader extends VirtualPage and is quite similar to it.
+
+Like with a VirtualPage - you should use `set()` to define content that will be loaded dynamically,
+while a spinner is shown to a user::
+
+    $loader = $app->add('Loader');
+    $loader->set(function($p) {
+
+        // Simulate slow-loading component
+        sleep(2);
+        $p->add('LoremIpsum');
+
+    });
+
+A good use-case example would be a dashboard graph. Unlike VirtualPage which is not visible to a regular render,
+Loader needs to occupy some space, so adding other objects inside content can be a good filler until content is loaded::
+
+    $loader = $app->add('Loader');
+    $loader->add(['Message', 'Please wait until we load LoremIpsum...', 'red']);
+    $loader->set(function($p) {
+
+        // Simulate slow-loading component
+        sleep(2);
+        $p->add('LoremIpsum');
+
+    });
+
```

Comment:
Do you think it is necessary to be able to remove the loading message once loader is finish? May be we can add this implementation inside Loader. 
or simply use code like this:
```
$loader = $app->add('Loader');
$m = $loader->add(['Message', 'Please wait until we load LoremIpsum...', 'red']);
$loader->needOnPageLoad = true;

$loader->set(function ($loader) use ($m) {
    //set your time expensive function here.
    sleep(2);
    $loader->add(new \atk4\ui\LoremIpsum(['size'=>1]));
    $m->js(true)->hide();
});
```

Reply:
I was thinking that the loading message would be appearing inside loader div and will be overwritten anyway with the desired content when loading is complete. 

As for the loader, if I understand correctly that's just a class which will be added/removed by JS, so we don't have to worry.